### PR TITLE
Timeline 2

### PR DESCRIPTION
**Pull request by taitus**
Timeline

Where
=====
* **Related Issue:** https://github.com/consul/consul/issues/1599
* **Related PR's:** https://github.com/rockandror/consul/issues/2

What
====
> Investments that have won during a participatory budget, go on to the next offline phases to become a reality.
> Add an admin section to update milestones and the corresponding user facing interface.

How
===
> Adding to `Investment` a has_many relation with new model `Milestone`
> Adding on `show` from `Admin::Budget::Investment` a new CRUD from `Milestone`
> Adding on Investment show view a new tab with milestones.

Screenshots
===========
> Front End
captura de pantalla 2017-06-23 a las 22 26 57

> Back End
captura de pantalla 2017-06-23 a las 22 30 27

Test
====
> Adding `Milestone` model and feature specs
> Increased the budget_investment spec feature to show milestone on tab

Deployment
==========
> As usual

Warnings
========
> None

### CODE REVIEW

File: app/views/budgets/investments/_milestones.html.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+          <%= t('budgets.investments.show.no_milestones') %>
+        </div>
+      <% end %>
+      <section class="timeline">
+        <ul class="no-bullet">
+          <% @investment.milestones.each do |milestone| %>
+            <li>
+              <div>
+                <h3><%= milestone.title %></h3>
+                <strong>Publicado el <%= milestone.created_at.strftime("%d/%m/%Y") %></strong>
```
Comment (bertocq) [inline app/views/budgets/investments/_milestones.html.erb]: I think the "Publicado el" should go as part of a translation literal, right? You can interpolate like:

`t("milestones.publish_date", publish_date:  milestone.created_at.strftime("%d/%m/%Y"))`

with `publish_date: "Publicado el %{publish_date}` on the .yml file

File: db/migrate/20170620132731_create_budget_investment_milestones.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class CreateBudgetInvestmentMilestones < ActiveRecord::Migration
+  def change
+    create_table :budget_investment_milestones do |t|
+      t.integer :investment_id
+      t.string   "title", limit: 80
```
Comment (bertocq) [inline db/migrate/20170620132731_create_budget_investment_milestones.rb]: We're already checking the title's presence on the model, but I'm always a fan of also setting database restrictions ( a `null: false` in this case) but no worries, can be added later on or even we can live without it I think

File: db/schema.rb [hunk 3]
```diff
@@ -924,7 +932,7 @@
     t.boolean  "email_digest",                              default: true
     t.boolean  "email_on_direct_message",                   default: true
     t.boolean  "official_position_badge",                   default: false
-    t.datetime "password_changed_at",                       default: '2016-12-21 17:55:08', null: false
+    t.datetime "password_changed_at",                       default: '2017-06-22 11:21:30', null: false
```
Comment (bertocq) [inline db/schema.rb]: Don't worry about this, I've it on my radar and will fix soon

File: spec/features/admin/budget_investments_spec.rb [hunk 4]
```diff
@@ -27,7 +27,7 @@
 
   context "Index" do
 
-    scenario 'Displaying investmentss' do
+    scenario 'Displaying investments' do
```
Comment (bertocq) [inline spec/features/admin/budget_investments_spec.rb]: 👏  nice catch! :)

File: app/models/budget/investment/milestone.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class Budget
+  class Investment
+    class Milestone < ActiveRecord::Base
+      belongs_to :investment
+
+      validates :title, presence: true
+      validates :investment, presence: true
+
+      def self.title_max_length
```
Comment (bertocq) [inline app/models/budget/investment/milestone.rb]: Allright, talked about it, we'll catch up with this later on along with budget model usage of similar code https://github.com/consul/consul/blob/master/app/models/budget.rb#L34-L40

File: app/views/budgets/investments/_comments.html.erb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<% commentable = comment_tree.commentable %>
+<% current_order = comment_tree.order %>
+
+<% cache [locale_and_user_status, current_order, commentable_cache_key(commentable), comment_tree.comments, comment_tree.comment_authors, commentable.comments_count, comment_flags] do %>
```
Comment (bertocq) [inline app/views/budgets/investments/_comments.html.erb]: Allright! Realized this is as well on budgets and comments partials... we'll deal with all once this goes into master 👍

**Comment by decabeza** [comment]
@taitus I think this PR it's perfect and ready to merge! 👏

Some minor details for future PR's (Don't worry now about this ones, I change it after in other mini PR) 😉 

- On `admin...\budget_investment_milestones/new` add `rows: 5` on textarea (a little more comfy to write)

- On budgets/investments show, the date info `Published 26/06/2017` have a low color contrast (my fault, I gonna add a Color Contrast Checker link on README files, WCAG AA is mandatory for correct accessibility)

- Personally I am not a fan of using `div` on css (there is a lot of them in HTML, its easy confused, I would change for example:

    - `tab-milestones ul div` to `tab-milestones ul .milestone-content`

**Comment by decabeza** [comment]
@taitus more minor comments 😄 (some only personal preferences)

- For create circles with **border-radius** I prefer use `rem-calc(same-width-and-height-of-element)` than **N%**.

- For elements with `position: absolute` it's ok use absolute **px**.

- For maintain vertical and horizontal rhythm there is a magic-css-trick, use always sizes relatives to **line-height** (24px in this case), for example:

    - Instead of padding: `rem-calc(3) rem-calc(10);`
    - Use: padding: `$line-height / 6 $line-height / 2;` (padding: 4px 12px)

- For **font-size**, it's better use Foundation `rem-calc` function instead of **rem units**.

- For last (this is not mandatory at all) I always write CSS attributes on alphabetical order (for me it's more fast to review). 🤓

**Comment by taitus** [comment]
@decabeza Perfect, I will comment with @agileontheweb and we will keep that in mind for the next commits 👍

**Pull request merged**